src/migrations/20230822192328-internal-goal-duplicate-cleanup.js

```diff
@@ -250,6 +259,211 @@ module.exports = {
+            ARRAY_AGG(DISTINCT "gr".id ORDER by "gr".id) "toRemove",
+            (ARRAY_AGG(DISTINCT "gr2".id))[1] "toUpdate"
+            FROM "GoalResources" "gr"
+            JOIN gtmm_recast gtmm
+            ON "gr"."goalId" = to_remove
+            LEFT JOIN "GoalResources" "gr2"
+            ON "gr2"."goalId" = gtmm."toUpdate"
+            AND "gr"."resourceId" = "gr2"."resourceId"
+            GROUP BY 1,2
+        );
+        SELECT * FROM "GoalResourcesToModify";
+        -- SELECT "resourceId", "toRemove", "toUpdate", COUNT(*), array_to_json(array_agg(row_to_json(gfmm)))
+        -- FROM "GoalResourcesToModify" gfmm
+        -- GROUP BY "resourceId", "toRemove", "toUpdate"
+        -- HAVING COUNT(*) > 1;
+
+        BEGIN;
+        SELECT set_config('audit.auditDescriptor', 'dup_goals_Insert_GoalResources', TRUE) as "auditDescriptor";
+        DROP TABLE IF EXISTS "InsertGoalResources";
+        CREATE TEMP TABLE "InsertGoalResources" AS
+            WITH  insert_goal_resources AS (
+                INSERT INTO "GoalResources"
+                (
+                "goalId",
+                "resourceId",
+                "sourceFields",
+                "createdAt",
+                "updatedAt",
+                "onAR",
+                "onApprovedAR"
+                )
+                SELECT
+                "goalId",
+                "resourceId",
+                "sourceFields",
+                "createdAt",
+                "updatedAt",
+                "onAR",
+                "onApprovedAR"
+                FROM "GoalResourcesToModify" grtm
+                WHERE grtm."toUpdate" IS NULL and grtm."goalId" IS NOT NULL
+                RETURNING
+                id "goalResourceId",
+                "goalId"
+            )
+        SELECT * FROM insert_goal_resources;
+        SELECT * FROM "InsertGoalResources";
+        END;
+
+        BEGIN;
+        SELECT set_config('audit.auditDescriptor', 'dup_goals_Update_GoalResources', TRUE) as "auditDescriptor";
+        DROP TABLE IF EXISTS "UpdateGoalResources";
+        CREATE TEMP TABLE "UpdateGoalResources" AS
+            WITH update_goal_resources AS (
+                UPDATE "GoalResources" "gr"
+                SET
+                "sourceFields" = grtm."sourceFields",
+                "createdAt" = grtm."createdAt",
+                "updatedAt" = grtm."updatedAt",
+                "onAR" = grtm."onAR",
+                "onApprovedAR" = grtm."onApprovedAR"
+                FROM "GoalResourcesToModify" grtm
+                WHERE "gr".id = grtm."toUpdate"
+                AND (
+                    "gr"."sourceFields" != grtm."sourceFields"
+                    OR "gr"."createdAt" != grtm."createdAt"
+                    OR "gr"."updatedAt" != grtm."updatedAt"
+                    OR "gr"."onAR" != grtm."onAR"
+                    OR "gr"."onApprovedAR" != grtm."onApprovedAR"
+                )
```

Comment:
I think this is a little more readable. There's also a form that uses hashed concatenations, but you wouldn't want to be doing that in a join like this and we don't need it for speed.
```suggestion
                AND NOT (
                    "gr"."sourceFields" = grtm."sourceFields"
                    AND "gr"."createdAt" = grtm."createdAt"
                    AND "gr"."updatedAt" = grtm."updatedAt"
                    AND "gr"."onAR" = grtm."onAR"
                    AND "gr"."onApprovedAR" = grtm."onApprovedAR"
                )
```

Reply:
I occurs to me that both of these are vulnerable to one of the fields being null and the join dropping. Not sure I want to bother coalescing if they're all NOT NULL fields.